File: Leetcode/0042-TrappingRainWater/java/app/src/main/java/TrappingRainWater/Solution2.java
package TrappingRainWater;

public class Solution2 {
    public int trap(int[] height) {
        int res = 0;
        int left = 0, right = height.length - 1;
        int leftMax = height[left], rightMax = height[right];
        while (left <= right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (leftMax < rightMax) {
                // leftMax determines the rain amount stored at location "left"
                res += leftMax - height[left];
                left++;
            } else {
                res += rightMax - height[right];
                right--;
            }
        }
        return res;
    }
}

File: Leetcode/0042-TrappingRainWater/go/solution2.go
func trap(height []int) int {
    n := len(height)
    // leftMax[i]: max height in range [0, i]
    leftMax := make([]int, n)
    leftMax[0] = height[0]
    // rightMax[i]: max height in range [i, n)
    rightMax := make([]int, n)
    rightMax[n-1] = height[n-1]
    for i := 1; i < n; i++ {
        j := n-i-1
        leftMax[i] = max(leftMax[i-1], height[i])
        rightMax[j] = max(rightMax[j+1], height[j])
    }

    res := 0
    for i := 1; i < n-1; i++ {
        rain := min(leftMax[i], rightMax[i])-height[i]
        res += rain
    }
    return res
}